context/src/main/java/io/opentelemetry/context/ThreadLocalContextStorage.java

```diff
@@ -30,14 +31,30 @@ public Scope attach(Context toAttach) {
 
     THREAD_LOCAL_STORAGE.set(toAttach);
 
-    return () -> {
-      if (current() != toAttach) {
+    return new ScopeImpl(beforeAttach, toAttach);
+  }
+
+  private class ScopeImpl implements Scope {
+    @Nullable private final Context beforeAttach;
+    private final Context toAttach;
+    private boolean closed;
+
+    private ScopeImpl(@Nullable Context beforeAttach, @Nonnull Context toAttach) {
```

Comment:
`Nonnull` isn't needed because arguments are assumed to be nonnull unless indicated with `@Nullable`.